Fix meaning of MasterConns metric

KJTsanaktsidis opened the pull request:
While exploring some of the mgo metrics around connections opening/closing, I noticed that the MasterConns metric, which I assume is _supposed_ to be an always-incrementing counter, actually gets decremented in the connection pool shrinking logic.

I assume that this is a mistake because other places that can reduce the count of open master connections, such as `kill()` in `abend` error scenarios, don't also decrease this number.

To replace the change in MasterConns metric that the pool shrinker was previously doing, I added its own counter to measure what's happening there (ShrunkConns).

I also added a counter for connections that failed abnormally (FailedConns). This is data that I'm quite curious to get my hands on to debug our Mongo deployment because we're seeing a scenario where there are a lot of connection failures in our setup and a huge churn in connections getting opened/closed.

Commit:
[OUT-5021] Ensure that MasterConns stat is a counter

I believe the MasterConns stat is supposed to be a counter that is never
decrementing. However, it actually does get decremented in the pool
shrinker.

Having it be an always-incrementing counter means you can use your
monitoring to do things like measure the rate at which connections are
opened.

If it was supposed to be decremented on connection close, it should be
done in kill() where it would also happen on connection failure or
abnormal connection close.

Commit:
[OUT-5021] Add metrics for shrunk and failed connections

Since I took away the change to the MasterConns and SlaveConns metrics
in the pool shrinker, I replaced it with a metric to capture when the
pool shrinking happens.

I also added a metric incrmeneted when a connection is abended. Frequent
opening and closing of connections is a sign of a serious problem with
Mongo and it's something I'd like to keep tabs on.

eminano committed:
Merge branch 'development' into ktsanaktsidis/OUT-5021/ensure_conns_is_counter

eminano commented on socket.go:
```diff
@@ -384,6 +384,9 @@ func (socket *mongoSocket) kill(err error, abend bool) {
 	socket.conn.Close()
 	stats.socketsAlive(-1)
+	if abend {
+		stats.failedConn(1)
```
Any reason not to do this at the end of the function in the existing abend check?

KJTsanaktsidis replied:
Mostly because I thought it would make more sense to increment this together with the `socketsAlive` metric. I don't think there's any technical reason it has to be here.

eminano commented on server.go:
```diff
@@ -530,7 +530,7 @@ func (server *mongoServer) poolShrinker() {
 			}
 			server.liveSockets = remainSockets
-			stats.conn(-1*end, server.info.Master)
+			stats.shrunkConn(1)
```
Shouldn't this be ```end``` instead of 1?

KJTsanaktsidis replied:
Yeah this is definitely a mistake, I will fix this.

eminano commented:
@KJTsanaktsidis, thanks for taking the time to open this PR and sorry for the delay!

Cheers,
Esther

Commit:
[OUT-5021] Record the actual number of shrunk conns in pool shrinker

The pool shrinker metric should record the number of connections that it
closed, not the number of times that it closed connections.

eminano committed:
Merge branch 'development' into ktsanaktsidis/OUT-5021/ensure_conns_is_counter
